src/iperf_client_api.c

```diff
@@ -580,8 +583,9 @@ iperf_run_client(struct iperf_test * test)
 
     /* Begin calculating CPU utilization */
     cpu_util(NULL);
+    rcv_timeout_value_in_us = (test->settings->rcv_timeout.secs * SEC_TO_US) + test->settings->rcv_timeout.usecs;
```

Comment:
This change creates a scenario with an early timeout. 
Given:
- `test->mode == SENDER`
- `rcv_timeout_value_in_us > 0`
- `test->state == TEST_END` (or `EXCHANGE_RESULTS` or `DISPLAY_RESULTS`)

This implies that `rcv_timeout_us = 0`. Then on line 642, the if statement will evaluate to something like `if (t_usecs > 0)`; always being true since `t_usecs` will pretty much always be greater than 0.

It might make more sense to split the blocks so the correct timeout is used:
(I've also renamed `rcv_timeout_value_in_us` -> `end_rcv_timeout` and `rcv_timeout_us` -> `running_rcv_timeout` to hopefully make their use more clear)
```C
    if (result < 0 && errno != EINTR) {
  	    i_errno = IESELECT;
	    goto cleanup_and_fail;
    } else if ( result == 0 && (running_rcv_timeout > 0 && test->state == TEST_RUNNING)) {
        /*
            * If nothing was received in non-reverse running state
            * then probably something got stuck - either client,
            * server or network, and test should be terminated./
            */
        iperf_time_now(&now);
        if (iperf_time_diff(&now, &last_receive_time, &diff_time) == 0) {
            t_usecs = iperf_time_in_usecs(&diff_time);
            if (t_usecs > running_rcv_timeout) {
                /* Idle timeout if no new blocks received */
                if (test->blocks_received == last_receive_blocks) {
                    i_errno = IENOMSG;
                    goto cleanup_and_fail;
                }
            }

        }
    } else if (result == 0 && (end_rcv_timeout > 0 && (test->state == TEST_END 
                                        || test->state == EXCHANGE_RESULTS 
                                        || test->state == DISPLAY_RESULTS))) {
        iperf_time_now(&now);
        if (iperf_time_diff(&now, &last_receive_time, &diff_time) == 0) {
            t_usecs = iperf_time_in_usecs(&diff_time);
            if (t_usecs > end_rcv_timeout) {
                /* Idle timeout if no new blocks received */
                if (test->blocks_received == last_receive_blocks) {
                    i_errno = IENOMSG;
                    goto cleanup_and_fail;
                }
            }

        }
    }
```

Reply:
Only `rcv_timeout_us` value is used for timeout (in the `select` statement).  The value of `rcv_timeout_value_in_us` is only used to initialize `rcv_timeout_us` and to **indicate** for for the ending states whether receive timeout was requested (as there is no "timeout requested" setting).  Therefore, the current code is correct in that respect.

I agree that probably `rcv_timeout_value_in_us` may have a better name, like `rcv_timeout_setting_us` but I won't make a change just for that.